File: 3-cloudrun-golang/repository/device_repository.go
package repository

import (
	"strings"
	"sync"

	"github.com/benchmark/cloudrun-go/model"
)

type DeviceRepository struct {
	mu    sync.RWMutex
	store map[string]model.Device
}

func NewDeviceRepository() *DeviceRepository {
	return &DeviceRepository{
		store: make(map[string]model.Device),
	}
}

func (r *DeviceRepository) Save(device model.Device) model.Device {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.store[device.ID] = device
	return device
}

func (r *DeviceRepository) FindByID(id string) (model.Device, bool) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	device, ok := r.store[id]
	return device, ok
}

func (r *DeviceRepository) FindByStatus(status string) []model.Device {
	r.mu.RLock()
	defer r.mu.RUnlock()

	var result []model.Device
	for _, d := range r.store {
		if strings.EqualFold(d.Status, status) {
			result = append(result, d)
		}
	}
	return result
}

func (r *DeviceRepository) FindAll() []model.Device {
	r.mu.RLock()
	defer r.mu.RUnlock()

	result := make([]model.Device, 0, len(r.store))
	for _, d := range r.store {
		result = append(result, d)
	}
	return result
}

File: 1-gke-quarkus-jvm/src/main/java/com/benchmark/device/repository/DeviceRepository.java
package com.benchmark.device.repository;

import com.benchmark.device.model.Device;
import jakarta.enterprise.context.ApplicationScoped;

import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

@ApplicationScoped
public class DeviceRepository {

    private final ConcurrentHashMap<UUID, Device> store = new ConcurrentHashMap<>();

    public Device save(Device device) {
        // Prevent OOM during heavy load testing
        if (store.size() > 1000) {
            store.remove(store.keys().nextElement());
        }
        store.put(device.getId(), device);
        return device;
    }

    public Optional<Device> findById(UUID id) {
        return Optional.ofNullable(store.get(id));
    }

    public List<Device> findByStatus(String status) {
        return store.values().stream()
                .filter(d -> d.getStatus().equalsIgnoreCase(status))
                .toList();
    }

    public List<Device> findAll() {
        return List.copyOf(store.values());
    }
}
